Patches/CheckGameEndPatch.cs

```diff
@@ -492,61 +492,28 @@ public bool CheckGameEndByLivingPlayers(out GameOverReason reason)
-            int Med = Utils.AlivePlayersCount(CountTypes.Medusa);
-            int SC = Utils.AlivePlayersCount(CountTypes.Spiritcaller);
-            int Glitch = Utils.AlivePlayersCount(CountTypes.Glitch);
-            int Arso = Utils.AlivePlayersCount(CountTypes.Arsonist);
-            int Shr = Utils.AlivePlayersCount(CountTypes.Shroud);
-            int WW = Utils.AlivePlayersCount(CountTypes.Werewolf);
-            int RR = Utils.AlivePlayersCount(CountTypes.RuthlessRomantic);
-
-            Imp += Main.AllAlivePlayerControls.Count(x => x.GetCustomRole().IsImpostor() && x.Is(CustomRoles.DualPersonality));
-            Crew += Main.AllAlivePlayerControls.Count(x => x.GetCustomRole().IsCrewmate() && x.Is(CustomRoles.DualPersonality));
-            Crew += Main.AllAlivePlayerControls.Count(x => (x.GetCustomRole().IsImpostor() && x.Is(CustomRoles.Admired)) && x.Is(CustomRoles.DualPersonality));
-            Crew += Main.AllAlivePlayerControls.Count(x => (x.GetCustomRole().IsNeutral() && x.Is(CustomRoles.Admired)) && x.Is(CustomRoles.DualPersonality));
-            Crew += Main.AllAlivePlayerControls.Count(x => (x.GetCustomRole().IsCrewmate() && x.Is(CustomRoles.Admired)) && x.Is(CustomRoles.DualPersonality));
-            CM += Main.AllAlivePlayerControls.Count(x => x.Is(CustomRoles.Charmed) && x.Is(CustomRoles.DualPersonality));
-            Jackal += Main.AllAlivePlayerControls.Count(x => x.Is(CustomRoles.Sidekick) && x.Is(CustomRoles.DualPersonality));
-            Jackal += Main.AllAlivePlayerControls.Count(x => x.Is(CustomRoles.Recruit) && x.Is(CustomRoles.DualPersonality));
-            Vamp += Main.AllAlivePlayerControls.Count(x => x.Is(CustomRoles.Infected) && x.Is(CustomRoles.DualPersonality));
-            Virus += Main.AllAlivePlayerControls.Count(x => x.Is(CustomRoles.Contagious) && x.Is(CustomRoles.DualPersonality));
-            Imp += Main.AllAlivePlayerControls.Count(x => x.Is(CustomRoles.Madmate) && x.Is(CustomRoles.DualPersonality));
-
-            int totalNKAlive = new int[] { Rit, Traitor, Med, PP, Jackal, Vamp, DH, Rogue, Wraith, Agitater, Pestilence, PB, Juggy, Doppelganger, Hunt, Necro, Pyro, Hex, Bandit, RR, WW, Shr, Arso, Glitch, Jinx, SK, Pel, Gam, BK, Pois, Virus, SC, CM }.Sum(); // Occ,
-
-            CustomWinner? winner = null;
-            CustomRoles? rl = null;
+            for (int i = 0; i < apcList.Count; i++)
+            {
+                PlayerControl pc = apcList[i];
+                if (pc == null) continue;
+
+                dual = pc.Is(CustomRoles.DualPersonality) ? 1 : 0;
+                var countType = Main.PlayerStates[pc.PlayerId].countTypes;
+                if (countType == CountTypes.OutOfGame || countType == CountTypes.None || countType == CountTypes.NWitch) continue;
+
+                if (roleCounts.ContainsKey(countType))
+                    roleCounts[countType]++;
+                else
+                    roleCounts[countType] = 1;
```

Comment:
I'm tired so I may be wrong but.... if the dictionary doesn't contain the key (goes to else) and you still try to read that key, won't it just throw a KeyNotFoundException?
You'd need to add the key with a value of 1 to the dictionary using .Add or .TryAdd

Reply:
Nope, you can add a key this way too